File: src/keboola_mcp_server/links.py
from typing import Literal

from pydantic import BaseModel, Field

from keboola_mcp_server.client import KeboolaClient

URLType = Literal['ui-detail', 'ui-dashboard', 'docs']


class Link(BaseModel):
    type: URLType = Field(..., description='The type of the URL.')
    title: str = Field(..., description='The name of the URL.')
    url: str = Field(..., description='The URL.')

    @classmethod
    def detail(cls, label: str, name: str, url: str) -> 'Link':
        return cls(type='ui-detail', title=f'{label}: {name}', url=url)

    @classmethod
    def dashboard(cls, title: str, url: str) -> 'Link':
        return cls(type='ui-dashboard', title=title, url=url)

    @classmethod
    def docs(cls, title: str, url: str) -> 'Link':
        return cls(type='docs', title=title, url=url)


class ProjectLinksManager:

    FLOW_DOCUMENTATION_URL = 'https://help.keboola.com/flows/'

    def __init__(self, base_url: str, project_id: str):
        self.base_url = base_url
        self.project_id = project_id

    @classmethod
    async def from_client(cls, client: KeboolaClient) -> 'ProjectLinksManager':
        base_url = client.storage_client.base_api_url
        project_id = await client.storage_client.project_id()
        return ProjectLinksManager(base_url, project_id)

    def get_component_configuration_url(self, component_id: str, configuration_id: str) -> str:
        """Get the UI detail URL for a specific component configuration."""
        return f'{self.base_url}/admin/projects/{self.project_id}/components/{component_id}/{configuration_id}'

    def get_component_configurations_dashboard_url(self, component_id: str) -> str:
        """Get the UI dashboard URL for all configurations of a component."""
        return f'{self.base_url}/admin/projects/{self.project_id}/components/{component_id}'

    def get_flow_url(self, flow_id: str | int) -> str:
        """Get the UI detail URL for a specific flow."""
        return f'{self.base_url}/admin/projects/{self.project_id}/flows/{flow_id}'

    def get_flows_dashboard_url(self) -> str:
        """Get the UI dashboard URL for all flows in a project."""
        return f'{self.base_url}/admin/projects/{self.project_id}/flows'

    def get_job_url(self, job_id: str) -> str:
        """Get the UI detail URL for a specific job."""
        return f'{self.base_url}/admin/projects/{self.project_id}/queue/{job_id}'

    def get_jobs_dashboard_url(self) -> str:
        """Get the UI dashboard URL for all jobs in a project."""
        return f'{self.base_url}/admin/projects/{self.project_id}/queue'

    def get_bucket_url(self, bucket_id: str) -> str:
        """Get the UI detail URL for a specific bucket."""
        return f'{self.base_url}/admin/projects/{self.project_id}/storage/{bucket_id}'

    def get_buckets_dashboard_url(self) -> str:
        """Get the UI dashboard URL for all buckets in a project."""
        return f'{self.base_url}/admin/projects/{self.project_id}/storage'

    def get_project_url(self) -> str:
        """Return the UI URL for accessing the project."""
        return f'{self.base_url}/admin/projects/{self.project_id}'

    def get_table_url(self, bucket_id: str, table_name: str) -> str:
        """Return the UI URL for accessing the table of a bucket."""
        return f'{self.base_url}/admin/projects/{self.project_id}/storage/{bucket_id}/table/{table_name}'

    def get_project_links(self) -> list[Link]:
        return [Link.detail(label='Project Dashboard', name='', url=self.get_project_url())]

    def get_flow_links(self, flow_id: str | int, flow_name: str) -> list[Link]:
        return [
            Link.detail(label='Flow', name=flow_name, url=self.get_flow_url(flow_id)),
            Link.dashboard(title='Flows in the project', url=self.get_flows_dashboard_url()),
            Link.docs(title='Documentation for Keboola Flows', url=self.FLOW_DOCUMENTATION_URL),
        ]

    def get_component_configuration_links(
        self, component_id: str, configuration_id: str, configuration_name: str
    ) -> list[Link]:
        return [
            Link.detail(
                label='Configuration',
                name=configuration_name,
                url=self.get_component_configuration_url(component_id, configuration_id),
            ),
            Link.dashboard(
                title='Component Configurations Dashboard',
                url=self.get_component_configurations_dashboard_url(component_id),
            ),
        ]

    def get_job_links(self, job_id: str) -> list[Link]:
        return [
            Link.detail(label='Job', name=job_id, url=self.get_job_url(job_id)),
            Link.dashboard(title='Jobs Dashboard', url=self.get_jobs_dashboard_url()),
        ]

    def get_bucket_links(self, bucket_id: str, bucket_name: str) -> list[Link]:
        return [
            Link.detail(label='Bucket', name=bucket_name, url=self.get_bucket_url(bucket_id)),
            Link.dashboard(title='Buckets in the project', url=self.get_buckets_dashboard_url()),
        ]

    def get_table_links(self, bucket_id: str, table_name: str) -> list[Link]:
        return [
            Link.detail(label='Table', name=table_name, url=self.get_table_url(bucket_id, table_name)),
        ]

File: src/links.ts
import {
  CONDITIONAL_FLOW_COMPONENT_ID,
  DATA_APP_COMPONENT_ID,
  FLOW_TYPES,
  type FlowType,
} from '@/constants';

/**
 * UI / docs links surfaced to the user alongside tool results. Faithful port of
 * the Python `links.ProjectLinksManager` — pure URL building from the project's
 * base URL, project id, and (optional) dev-branch id.
 */
export type UrlType = 'ui-detail' | 'ui-dashboard' | 'docs';

export type Link = {
  type: UrlType;
  title: string;
  url: string;
};

const detail = (title: string, url: string): Link => ({ type: 'ui-detail', title, url });
const dashboard = (title: string, url: string): Link => ({ type: 'ui-dashboard', title, url });
const docs = (title: string, url: string): Link => ({ type: 'docs', title, url });

const FLOW_DOCUMENTATION_URL = 'https://help.keboola.com/flows/';

const isFlowType = (componentId: string | undefined): componentId is FlowType =>
  componentId !== undefined && (FLOW_TYPES as readonly string[]).includes(componentId);

const isDataAppComponent = (componentId: string | undefined): boolean =>
  componentId === DATA_APP_COMPONENT_ID;

const isTransformationComponent = (componentId: string): boolean =>
  Boolean(componentId) && componentId.includes('transformation');

export class ProjectLinksManager {
  private readonly baseUrl: string;
  private readonly projectId: string;
  private readonly branchId: string | undefined;

  constructor(options: { baseUrl: string; projectId: string; branchId?: string }) {
    this.baseUrl = options.baseUrl;
    this.projectId = options.projectId;
    this.branchId = options.branchId;
  }

  private url(path: string): string {
    const parts = [this.baseUrl, 'admin/projects', this.projectId];
    if (this.branchId) {
      parts.push('branch', this.branchId);
    }
    parts.push(path);
    return parts.join('/');
  }

  private flowPath(flowType: FlowType): string {
    return flowType === CONDITIONAL_FLOW_COMPONENT_ID ? 'flows-v2' : 'flows';
  }

  /** Most relevant links for a Keboola object from mutually-exclusive identifiers. */
  getLinks(opts: {
    bucketId?: string;
    tableId?: string;
    componentId?: string;
    configurationId?: string;
    name?: string;
  }): Link[] {
    const { bucketId, tableId, componentId, configurationId, name } = opts;
    if (componentId && configurationId) {
      return [this.getComponentConfigLink(componentId, configurationId, name ?? '')];
    }
    if (componentId) {
      return [this.getConfigDashboardLink(componentId, name ?? '')];
    }
    if (tableId) {
      return [this.getTableDetailLinkFromTableId(tableId)];
    }
    if (bucketId) {
      return [this.getBucketDetailLink(bucketId, name ?? bucketId)];
    }
    return [];
  }

  // --- Project ---
  getProjectDetailLink(): Link {
    return detail('Project Dashboard', this.url(''));
  }

  getProjectLinks(): Link[] {
    return [this.getProjectDetailLink()];
  }

  // --- Flows ---
  getFlowDetailLink(flowId: string | number, flowName: string, flowType: FlowType): Link {
    return detail(`Flow: ${flowName}`, this.url(`${this.flowPath(flowType)}/${flowId}`));
  }

  getFlowsDashboardLink(flowType: FlowType): Link {
    const label = flowType === CONDITIONAL_FLOW_COMPONENT_ID ? 'Conditional Flows' : 'Flows';
    return dashboard(`${label} in the project`, this.url(this.flowPath(flowType)));
  }

  getFlowsDocsLink(): Link {
    return docs('Documentation for Keboola Flows', FLOW_DOCUMENTATION_URL);
  }

  getFlowLinks(flowId: string | number, flowName: string, flowType: FlowType): Link[] {
    return [
      this.getFlowDetailLink(flowId, flowName, flowType),
      this.getFlowsDashboardLink(flowType),
      this.getFlowsDocsLink(),
    ];
  }

  // --- Schedulers ---
  getSchedulerDetailLink(flowId: string | number, flowType: FlowType): Link {
    return detail('Schedules', this.url(`${this.flowPath(flowType)}/${flowId}/schedules`));
  }

  // --- Components ---
  getComponentConfigLink(
    componentId: string,
    configurationId: string,
    configurationName: string,
  ): Link {
    if (isTransformationComponent(componentId)) {
      return this.getTransformationConfigLink(componentId, configurationId, configurationName);
    }
    if (isDataAppComponent(componentId)) {
      return this.getDataAppConfigLink(configurationId, configurationName, false);
    }
    if (isFlowType(componentId)) {
      return this.getFlowDetailLink(configurationId, configurationName, componentId);
    }
    return detail(
      `Configuration: ${configurationName}`,
      this.url(`components/${componentId}/${configurationId}`),
    );
  }

  getConfigDashboardLink(componentId: string, componentName: string | undefined): Link {
    const label = componentName ? componentName : `Component "${componentId}"`;
    return dashboard(`${label} Configurations Dashboard`, this.url(`components/${componentId}`));
  }

  getUsedComponentsLink(): Link {
    return dashboard('Used Components Dashboard', this.url('components/configurations'));
  }

  getConfigurationLinks(
    componentId: string,
    configurationId: string,
    configurationName: string,
  ): Link[] {
    return [
      this.getComponentConfigLink(componentId, configurationId, configurationName),
      this.getConfigDashboardLink(componentId, undefined),
    ];
  }

  // --- Data Apps ---
  getDataAppConfigLink(
    configurationId: string,
    configurationName: string,
    usesBasicAuthentication: boolean,
  ): Link {
    const title = usesBasicAuthentication
      ? `Data App Configuration (To see password, click on "OPEN DATA APP"): ${configurationName}`
      : `Data App Configuration: ${configurationName}`;
    return detail(title, this.url(`data-apps/${configurationId}`));
  }

  getDataAppDashboardLink(): Link {
    return dashboard('Data Apps in the project', this.url('data-apps'));
  }

  getDataAppDeploymentLink(deploymentLink: string): Link {
    return detail('Data App Deployment', deploymentLink);
  }

  getDataAppLinks(
    configurationId: string,
    configurationName: string,
    deploymentLink?: string,
    usesBasicAuthentication = false,
  ): Link[] {
    const links = [
      this.getDataAppConfigLink(configurationId, configurationName, usesBasicAuthentication),
      this.getDataAppDashboardLink(),
    ];
    if (deploymentLink) {
      links.push(this.getDataAppDeploymentLink(deploymentLink));
    }
    return links;
  }

  // --- Transformations ---
  getTransformationsDashboardLink(): Link {
    return dashboard('Transformations dashboard', this.url('transformations-v2'));
  }

  getTransformationConfigLink(
    transformationType: string,
    transformationId: string,
    transformationName: string,
  ): Link {
    return detail(
      `Transformation: ${transformationName}`,
      this.url(`transformations-v2/${transformationType}/${transformationId}`),
    );
  }

  getTransformationLinks(
    transformationType: string,
    transformationId: string,
    transformationName: string,
  ): Link[] {
    return [
      this.getTransformationConfigLink(transformationType, transformationId, transformationName),
      this.getTransformationsDashboardLink(),
    ];
  }

  // --- Jobs ---
  getJobDetailLink(jobId: string): Link {
    return detail(`Job: ${jobId}`, this.url(`queue/${jobId}`));
  }

  getJobsDashboardLink(): Link {
    return dashboard('Jobs in the project', this.url('queue'));
  }

  getJobLinks(jobId: string): Link[] {
    return [this.getJobDetailLink(jobId), this.getJobsDashboardLink()];
  }

  // --- Buckets ---
  getBucketDetailLink(bucketId: string, bucketName: string): Link {
    return detail(`Bucket: ${bucketName}`, this.url(`storage/${bucketId}`));
  }

  getBucketDashboardLink(): Link {
    return dashboard('Buckets in the project', this.url('storage'));
  }

  getBucketLinks(bucketId: string, bucketName: string): Link[] {
    return [this.getBucketDetailLink(bucketId, bucketName), this.getBucketDashboardLink()];
  }

  // --- Tables ---
  getTableDetailLink(bucketId: string, tableName: string): Link {
    return detail(`Table: ${tableName}`, this.url(`storage/${bucketId}/table/${tableName}`));
  }

  getTableDetailLinkFromTableId(tableId: string): Link {
    const segments = tableId.split('.');
    const tableName = segments[segments.length - 1]!;
    const bucketId = segments.slice(0, -1).join('.');
    return this.getTableDetailLink(bucketId, tableName);
  }

  getTableLinks(bucketId: string, bucketName: string, tableName: string): Link[] {
    return [
      this.getTableDetailLink(bucketId, tableName),
      this.getBucketDetailLink(bucketId, bucketName),
    ];
  }
}
